README
App-TimeClock

This is a simple reporting utility for timeclock, which is an Emacs
time tracking package.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc App::TimeClock

You can also look for information at:

    Issue tracker on Github (report bugs here)
        https://github.com/soren/App-TimeClock/issues

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/App-TimeClock

    CPAN Ratings
        http://cpanratings.perl.org/d/App-TimeClock

    Search CPAN
        http://search.cpan.org/dist/App-TimeClock/


LICENSE AND COPYRIGHT

Copyright (C) 2012-2018 Søren Lund

This file is part of App::TimeClock.

App::TimeClock is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

App::TimeClock is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with App::TimeClock.  If not, see <http://www.gnu.org/licenses/>.


Changes
Revision history for App-TimeClock

0.01    2012-03-15 22:20
        First version, released on an unsuspecting world.

0.02    2012-03-16 11:07
        Added description of .timelogrc configuration file.
        Fixed a few links in the documentation.

0.03    2012-03-16 16:02
        ConsolePrinter: Added a blank line before printing the header.
        timeclock.pl: Added --version option that will print the version number.
        Switched to GPL v3.
        Updated documentation.

0.04    2012-11-28 10:33
        ConsolePrinter: Fixed bug that would cause columns not to align when task descriptions contained utf8 characters.
        ConsolePrinter: Fixed divions by zero bug (if worked less that a day).
        Build.PL: Fixed license value in Build file.
        Added more test files.

0.05    2012-11-28 11:43
        DailyReport: The timelog file is now read as utf8

0.06    2012-12-11 09:42
        DailyReport: Now shows current (or dangling) projects and projects crossing midnight.
        Added tests of the console reports.

0.07    2012-12-13 10:58
        Removed dependency on Time::Piece, making us compatible with older versions of Perl.
        Updated to GPL v3 in source files. 
        Fixed a few typos in the documentation.

0.08    2013-03-09 21:32
        Version and copyright information now prints correctly using UTF-8.
        Now supports DOS formatted timelogs (CR+NL) even on Linux/Unix and vice versa.
        Changed from RT to GitHub for issue tracking.

0.09   2013-03-11 08:53
       Included test file which was forgotten in previous release.
       Updated copyright year in main documentation.

0.10   2015-09-07 22:21
       Fixed Issue #13 by accepting all codes: b, h, i, o and O.
       Note that codes b and h will just be skipped and O will be treated as o.

0.11   2016-12-30 10:18
       Fixed Issue #16 by supporting new default location of the timelog file in Emacs 25.

0.12   2017-01-02 10:28
       Fixed Issue #17 concerning unit tests failing when locale is not en_US.UTF-8

0.13   2017-01-04 11:14
       Fixed Issue #18 by implementing support for perl version pre 5.10. Also all CPAN Testers tests sould be green.
       Fixed Issue #20 by adding use strict/warnings, this should remove all Kwalitee issues.

0.14   2018-12-17 11:43
       Fixed Issue #22 and #23 by updating README and POD
       Added App Veyor CI configuration for testing on Windows.
       Updated Build badges.
       Fixed test on Haiko

       

lib/App/TimeClock/Daily/ConsolePrinter.pm
package App::TimeClock::Daily::ConsolePrinter;

use strict;
use warnings;

our @ISA = qw(App::TimeClock::Daily::PrinterInterface);

use POSIX qw(strftime);

use utf8;
binmode STDOUT, ':utf8';

our $hrline =  '+' . ('-' x 62) . '+' . ('-' x 7) . '+';

=head1 NAME

App::TimeClock::Daily::ConsolePrinter

=head1 DESCRIPTION

Implements the L<App::TimeClock::Daily::PrinterInterface>. Will print a simple ASCII
format. Suitable for using in a console/terminal.

=head1 METHODS

=over

=cut

=item print_header()

Prints a header including todays date. The header is indented to be
centered above the tables printed by L</print_day()>. Example:

                 =====================================
                 Daily Report Wed Mar 14 13:39:06 2012
                 =====================================

=cut
sub print_header {
    my $self = shift;
    my $ident = ' ' x 17;
    $self->_print("\n");
    $self->_print("${ident}=====================================\n");
    $self->_print("${ident}Daily Report " . localtime() . "\n");
    $self->_print("${ident}=====================================\n\n");
};

=item print_day()

Prints all activities for a day including a total. Is printed in a ACSII
table. Example:

 * Mon 2012/03/12 (08:21:16 - 16:05:31) *
 +--------------------------------------------------------------+-------+
 | Total Daily Hours                                            |  7.73 |
 +--------------------------------------------------------------+-------+
 | Lunch                                                        |  0.57 |
 +--------------------------------------------------------------+-------+
 | MyProject:Estimation                                         |  2.90 |
 +--------------------------------------------------------------+-------+
 | AnotherProject:BugFixing                                     |  4.26 |
 +--------------------------------------------------------------+-------+

=cut
sub print_day {
    my ($self, $date, $start, $end, $work, %projects) = (@_);
    my ($year, $mon, $mday) = split(/\//, $date);
    my $wday = substr(strftime("%a", 0, 0, 0, $mday, $mon-1, $year-1900),0,3);

    $self->_print(sprintf("* %3s %s (%s - %s) *\n", $wday, $date, $start, $end));
    $self->_print("$hrline\n");
    $self->_print(sprintf("| Total Daily Hours                                            | %5.2f |\n", $work));
    $self->_print("$hrline\n");

    foreach my $k (sort keys %projects) {
        $self->_print(sprintf("| %-60s | %5.2f |\n",$k, $projects{$k}));
        $self->_print("$hrline\n");
    }
    $self->_print("\n");
};

=item print_footer()

Prints the total number of hours worked and the daily
average. Example:

 TOTAL = 291.72 hours
 PERIOD = 42 days
 AVERAGE = 7.95 hours/day

=cut
sub print_footer {
    my ($self, $work_year_to_date, $day_count) = (@_);
    $self->_print(sprintf("TOTAL = %.2f hours\n", $work_year_to_date));
    $self->_print(sprintf("PERIOD = %d days\n", $day_count));
    $self->_print(sprintf("AVERAGE = %.2f hours/day\n", $day_count > 0 ? $work_year_to_date / $day_count : 0));
};
1;

=back

=for text
=encoding utf-8
=end

=head1 AUTHOR

Søren Lund, C<< <soren at lund.org> >>

=head1 SEE ALSO

L<timeclock.pl>

=head1 COPYRIGHT

Copyright (C) 2012-2018 Søren Lund

This file is part of App::TimeClock.

App::TimeClock is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

App::TimeClock is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with App::TimeClock.  If not, see <http://www.gnu.org/licenses/>.
